### ISSUE TITLE
Book metadata: new slot in sidecar ? and display enhancements ?

### ISSUE THREAD

**Issue by poire-z** [koreader/koreader#2986]
Found out a bug after I had disabled Reading Statistics (v1) plugin, when viewing Book Information from FileManager (hold on a file)
- if book never opened (no sidecar file), Book Information will open the document to extract the metadata.
- if book was opened, (sidecar file present), Book Information will use the "stats" slot from the sidecar file to get the metadata.  When Reading Statistics are enabled, it is filled with like:
```lua
    ["stats"] = {
        ["performance_in_pages"] = {
            [1498589145] = 8,
            [1498589129] = 7
        },
        ["authors"] = "Wikipedia FR",
        ["notes"] = 0,
        ["total_time_in_sec"] = 31,
        ["title"] = "Huns",
        ["highlights"] = 0,
        ["pages"] = 29,
        ["series"] = "",
        ["language"] = "fr"
    },
```
When Reading statistics are disabled, it is filled with:
```lua
    ["stats"] = {
        ["performance_in_pages"] = {},
        ["authors"] = "",
        ["total_time_in_sec"] = 0,
        ["notes"] = 0,
        ["title"] = "",
        ["highlights"] = 0,
        ["pages"] = 0,
        ["series"] = "",
        ["language"] = ""
    },
```
and we get a bunch of "N/A" in the Book Information page.
So, it's obvisouly a bug in the statistics.koplugin plugin, which we may fix there.
(I guess people never had the need to disable this plugin - I don't care much about statistics, but I never did disable it, except while playing a few weeks ago - but they may with the v2 sqlite based).

But I think we should find a better solution for that, knowing that:
- the Statistics plugin v2 (#2893) will use SQLite, and, having a quick glance at its code, will store title and authors into its sqlite db, and will stop filling this sidecar 'stats' slot (I don't think it should have to continue filling it). - and this plugin can be disabled.
- my CoverBrowser plugin may know about some book metadata, and I was thinking about how to have it pass them to the Book Information call, but it's nor obvious. It too can be disabled.

(Somehow related in #2905 (about using Book Status from FileManager): readerstatus.lua, as the book is already opened, will directly provided `props = self.document:getProps()` to the BookStatusWidget:new(). If it were to be used from File browser, it should find an other way to get these metadata.)

So, I think at least the core reader itself should now fill (on document opening, and update it on flushsettings?) a new slot with the info it has (now that the book is open), with at least:
```lua
    ["metadata"] = {
        ["authors"] = "Wikipedia FR",
        ["notes"] = 0, -- not sure about that, but it's reader stuff, so it may update that
        ["title"] = "Huns",
        ["highlights"] = 0, -- not sure about that, but it's reader stuff, so it may update that
        ["pages"] = 29, -- may need to be updated if one change font size, orientation...
        ["series"] = "",
        ["language"] = "fr"
    },
```
New "metadata" key ? or same 'stats' key, for existing sidecar files ?
(Side question: soon, the "description" metadata will be available, and it may be big (seen some 1-2 Kb text): can/should it be stored in the sidecar file too ?)

With that, books at least opened once with a Sidecar file not purged with have accurate information, even when all plugins disabled.
Optimaly, there should be a way for "metadata providers" (stats v2 plugin, which may have metadata even for purged sidecar files, or my coverbrowser plugin with its bookinfo db) to register themselves, so Book information can try to get metadata from them before opening the book (which may take time as we know). But that may get twisted and complex...

Also, talking about that here although it's a different problem, but it's related:
I hardly ever use Book Status (but I find it pretty), but I sometime use Book Information, and was thinking about adding Keywords and Description to it. It will not work as is, as these are long strings, and KeyValuePage won't wrap them, but truncate them:

<kbd> ![bookinformation](https://user-images.githubusercontent.com/24273478/27605960-92102e76-5b7e-11e7-974a-4eac4458fe63.png) </kbd>
So it may need a rewrite if we want to present them here.

I had a try at having description displayed in the onFileHold ButtonDialogTitle:

<kbd>![onfilehold](https://user-images.githubusercontent.com/24273478/27606032-d065ae12-5b7e-11e7-8bab-9159389689e5.png)</kbd>

which I'm not sure I like. But I think title and authors may have a place there below the filename if they are available (Book information was made a button I guess because it may be costly to get these metadata and it needs a bit of room to display some metadata).

Any thoughts ?

**Comment by Frenzie** [comment]
In principle I agree about the metadata thing, but since this is one of those things where the user won't normally mess around with the metadata files it's also defensible to say just stick with the "stats" that already exist (depending on whether there'd need to be any conversion/clean-up and such).

> So [KeyValuePage] may need a rewrite if we want to present them here.

Not necessarily. Changing the display would require a rewrite, but you could also add tapping on the truncated text.

Regarding the infobox I think your screenshot might look slightly crowded.

**Comment by poire-z** [comment]
> you could also add tapping on the truncated text.

Sounds nice, simple and generic enough, i'll have a go at adding that to `KeyValuePage`.
It already has a `tap` event registered if the caller provides a `callback` property to a `kv_pairs` item. But I see that currently, no caller provides one. Should I hijack `tap` event for displaying truncated key/value ? Or should I use `hold` for that ?

About the new Description metadata: from EPUBs, we often get HTML strings. Should I convert it to plain text early and store plain-text (in metadata.lua and bookinfo.db), so code using it don't have to ? Or should I store the HTML as-is , and lets code using it do the filtering ?
For now, I don't see where we could use HTML.
But storing HTML will allow us to fix/update the html to plain-text conversion (in case we need to adapt it) later without having to trash the stored entry (meaning purge .sdr or clean bookinfo db entry).

**Comment by Frenzie** [comment]
> Should I hijack tap event for displaying truncated key/value ? Or should I use hold for that ?

I forgot about that. Hold is probably better then.

> Should I convert it to plain text early and store plain-text (in metadata.lua and bookinfo.db), so code using it don't have to ? Or should I store the HTML as-is , and lets code using it do the filtering ?

I'm not entirely sure where the best place to do HTML conversion is, but I'd say the metadata isn't it for the reasons you stated. A rich text widget (probably using MuPDF, not cr, due to lower memory requirements) could take full advantage of the HTML in the future.

**Comment by poire-z** [comment]
OK about Hold and storing HTML.

Regarding https://github.com/koreader/koreader/issues/2905#issuecomment-308634249:
Shouldn't we add the directory and filename as two more key/value in Book information ? Before or after the other metadata ?
And shouldn't we add Book information to the reader hamburger menu (just below History and Book status), to allow, from the book reader itself, to see both the filename (above comment's request) and the new description (useful from inside the book, as, from what I've seen so far, it's often content not found in the book itself) ? Easier than adding them to Book status (which has a fixed layout and no room for other stuff).

**Comment by Frenzie** [comment]
Agreed on both.

**Comment by Frenzie** [comment]
Closed in #2993.
